### ISSUE TITLE
Incorrect type matching

### ISSUE THREAD

**Issue by nikfarce** [microsoft/TypeScript#51503]
# Bug Report

I apologize for my poor English - I use a translator.

Link to [CodeSandbox](https://codesandbox.io/s/ecstatic-panini-c44brd?file=/src/index.ts) (I could not set up TypeScript Playground)

### 🕗 Version & Regression Information

- TypeScript versions - 4.4.4 - 4.8.4 (Didn't check anymore)

### 💻 Code

```ts
import { ui } from "webix";

const btnConfig: ui.buttonConfig = {
  view: 'button',
  value: 'BUTTON',
};

const btn = <ui.button>ui(btnConfig);

function f1(btnConfig: ui.buttonConfig) {}

function f2(btn: ui.button) {}

f1(btnConfig); // OK
f1(btn); // OK ???

f2(btnConfig); // NO OK
f2(btn); // OK
```

### 🙁 Actual behavior

Hello! I found a bug when programming in the webix library. 
There are views and configs for them, in theory they are absolutely different types, but TypeScript can take one type for another type, and they are not always interchangeable.

### 🙂 Expected behavior

I expected ui.buttonConfig != ui.button anyway.

**Comment by MartinJohns** [comment]
`ui.buttonConfig` has **only** optional properties, which makes it a weak type. The type `ui.button` has all properties present which `ui.buttonConfig` **requires** (which are none), and none of the properties of `ui.button` is incompatible with `ui.buttonConfig`, so it's assignable.

This is working as intended.

**Comment by fatcerberus** [comment]
> The type `ui.button` has all properties present which `ui.buttonConfig` **requires** (which are none)

Hmm, do they have some properties in common?  [Weak type detection](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-4.html#weak-type-detection) should be kicking in if they don't.

**Comment by MartinJohns** [comment]
Yes, they have properties in common, at least `name` is present in both.
